File: backend/app/schemas/wine.py
"""Wine schemas."""

from datetime import date, datetime
from decimal import Decimal
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

from app.models.wine import WineType
from app.models.user_wine import WineStatus


class TasteProfile(BaseModel):
    """Wine taste profile."""

    body: int | None = Field(None, ge=1, le=5)
    tannin: int | None = Field(None, ge=1, le=5)
    acidity: int | None = Field(None, ge=1, le=5)
    sweetness: int | None = Field(None, ge=1, le=5)


class WineBase(BaseModel):
    """Base wine information."""

    name: str
    producer: str | None = None
    vintage: int | None = None
    grape_variety: list[str] | None = None
    region: str | None = None
    country: str | None = None
    appellation: str | None = None
    abv: Decimal | None = None
    type: WineType = WineType.RED


class WineResponse(WineBase):
    """Wine response model."""

    id: UUID
    taste_profile: TasteProfile | None = None
    food_pairing: list[str] | None = None
    flavor_notes: list[str] | None = None
    serving_temp_min: int | None = None
    serving_temp_max: int | None = None
    drinking_window_start: int | None = None
    drinking_window_end: int | None = None
    description: str | None = None
    image_url: str | None = None
    ai_confidence: Decimal | None = None

    model_config = ConfigDict(from_attributes=True)

    @classmethod
    def from_orm_with_profile(cls, wine):
        """Create response from ORM model with taste profile."""
        data = {
            "id": wine.id,
            "name": wine.name,
            "producer": wine.producer,
            "vintage": wine.vintage,
            "grape_variety": wine.grape_variety,
            "region": wine.region,
            "country": wine.country,
            "appellation": wine.appellation,
            "abv": wine.abv,
            "type": wine.type,
            "taste_profile": TasteProfile(
                body=wine.body,
                tannin=wine.tannin,
                acidity=wine.acidity,
                sweetness=wine.sweetness,
            ) if any([wine.body, wine.tannin, wine.acidity, wine.sweetness]) else None,
            "food_pairing": wine.food_pairing,
            "flavor_notes": wine.flavor_notes,
            "serving_temp_min": wine.serving_temp_min,
            "serving_temp_max": wine.serving_temp_max,
            "drinking_window_start": wine.drinking_window_start,
            "drinking_window_end": wine.drinking_window_end,
            "description": wine.description,
            "image_url": wine.image_url,
            "ai_confidence": wine.ai_confidence,
        }
        return cls(**data)


class TagInWine(BaseModel):
    """Tag info for wine response."""

    id: UUID
    name: str
    type: str
    color: str

    model_config = ConfigDict(from_attributes=True)


class UserWineCreate(BaseModel):
    """User wine creation request."""

    scan_id: str | None = None
    wine_id: UUID | None = None
    wine_overrides: WineBase | None = None
    quantity: int = Field(default=1, ge=1)
    purchase_date: date | None = None
    purchase_price: Decimal | None = Field(None, ge=0)
    purchase_place: str | None = None
    personal_note: str | None = None
    tag_ids: list[UUID] | None = None


class UserWineBatchCreate(BaseModel):
    """Batch user wine creation request."""

    scan_session_id: str
    wines: list[dict]  # List of wine creation data
    common_tags: list[UUID] | None = None
    common_purchase_date: date | None = None


class UserWineUpdate(BaseModel):
    """User wine update request."""

    quantity: int | None = Field(None, ge=0)
    purchase_date: date | None = None
    purchase_price: Decimal | None = Field(None, ge=0)
    purchase_place: str | None = None
    personal_note: str | None = None
    tag_ids: list[UUID] | None = None


class WineStatusUpdate(BaseModel):
    """Wine status update request."""

    status: Literal["consumed", "gifted"]
    quantity_change: int = Field(default=1, ge=1)
    consumed_date: date | None = None
    rating: int | None = Field(None, ge=1, le=5)
    tasting_note: str | None = None
    recipient: str | None = None  # For gifted


class WineQuantityUpdate(BaseModel):
    """Wine quantity update request."""

    action: Literal["increase", "decrease", "set"]
    amount: int = Field(..., ge=0)


class UserWineResponse(BaseModel):
    """User wine response model."""

    id: UUID
    quantity: int
    status: WineStatus
    purchase_date: date | None = None
    purchase_price: Decimal | None = None
    purchase_place: str | None = None
    personal_note: str | None = None
    personal_rating: int | None = None
    original_image_url: str | None = None
    created_at: datetime
    updated_at: datetime
    wine: WineResponse
    tags: list[TagInWine] = []
    drinking_status: str | None = None  # "aging", "optimal", "drink_soon", "urgent"

    model_config = ConfigDict(from_attributes=True)


class UserWineListResponse(BaseModel):
    """User wine list item (lighter response)."""

    id: UUID
    quantity: int
    status: WineStatus
    purchase_date: date | None = None
    purchase_price: Decimal | None = None
    created_at: datetime
    wine: WineResponse
    tags: list[TagInWine] = []
    drinking_status: str | None = None

    model_config = ConfigDict(from_attributes=True)


class WineFilterParams(BaseModel):
    """Wine filter parameters."""

    status: WineStatus | None = None
    type: WineType | None = None
    country: str | None = None
    grape: str | None = None
    tag_id: UUID | None = None
    drinking_window: Literal["now", "aging", "urgent"] | None = None
    min_price: int | None = None
    max_price: int | None = None
    search: str | None = None

File: frontend/src/types/wine.ts
// Wine Types

export type WineType = 'red' | 'white' | 'rose' | 'sparkling' | 'dessert' | 'fortified' | 'other';
export type WineStatus = 'owned' | 'consumed' | 'gifted';
export type DrinkingStatus = 'aging' | 'optimal' | 'drink_soon' | 'urgent';
export type DrinkingUrgency = 'drink_now' | 'drink_soon' | 'optimal' | 'can_wait';
export type TagType = 'cellar' | 'location' | 'custom';

export interface TasteProfile {
  body: number | null;
  tannin: number | null;
  acidity: number | null;
  sweetness: number | null;
}

export interface Wine {
  id: string;
  name: string;
  producer: string | null;
  vintage: number | null;
  grape_variety: string[] | null;
  region: string | null;
  country: string | null;
  appellation: string | null;
  abv: number | null;
  type: WineType;
  taste_profile: TasteProfile | null;
  food_pairing: string[] | null;
  flavor_notes: string[] | null;
  serving_temp_min: number | null;
  serving_temp_max: number | null;
  drinking_window_start: number | null;
  drinking_window_end: number | null;
  description: string | null;
  image_url: string | null;
  ai_confidence: number | null;
  ai_analysis: WineAIAnalysis | null;
}

export interface Tag {
  id: string;
  name: string;
  type: TagType;
  color: string;
  sort_order: number;
  wine_count: number;
  created_at: string;
}

export interface WineStatusHistory {
  id: string;
  status: 'consumed' | 'gifted';
  event_date: string;
  quantity: number;
  rating: number | null;
  note: string | null;
  recipient: string | null;
  created_at: string;
}

export interface UserWine {
  id: string;
  quantity: number;
  status: WineStatus;
  purchase_date: string | null;
  purchase_price: number | null;
  purchase_place: string | null;
  personal_note: string | null;
  personal_rating: number | null;
  original_image_url: string | null;
  label_number: string | null;
  created_at: string;
  updated_at: string;
  wine: Wine;
  tags: Tag[];
  status_histories: WineStatusHistory[];
  drinking_status: DrinkingStatus | null;
}

export interface UserWineListItem {
  id: string;
  quantity: number;
  status: WineStatus;
  purchase_date: string | null;
  purchase_price: number | null;
  label_number: string | null;
  created_at: string;
  wine: Wine;
  tags: Tag[];
  drinking_status: DrinkingStatus | null;
}

// Request types
export interface UserWineCreateRequest {
  scan_id?: string;
  wine_id?: string;
  wine_overrides?: Partial<Wine>;
  quantity?: number;
  purchase_date?: string;
  purchase_price?: number;
  purchase_place?: string;
  personal_note?: string;
  tag_ids?: string[];
}

export interface UserWineUpdateRequest {
  quantity?: number;
  purchase_date?: string;
  purchase_price?: number;
  purchase_place?: string;
  personal_note?: string;
  tag_ids?: string[];
}

export interface WineStatusUpdateRequest {
  status: 'consumed' | 'gifted';
  quantity_change?: number;
  consumed_date?: string;
  gifted_date?: string;
  rating?: number;
  tasting_note?: string;
  recipient?: string;
}

export interface WineQuantityUpdateRequest {
  action: 'increase' | 'decrease' | 'set';
  amount: number;
}

// Filter types
export interface WineFilterParams {
  status?: WineStatus;
  type?: WineType;
  country?: string;
  grape?: string;
  tag_id?: string;
  tag_ids?: string;
  drinking_window?: 'now' | 'aging' | 'urgent';
  min_price?: number;
  max_price?: number;
  search?: string;
}

// Scan types
export interface ScannedWine {
  name: string;
  producer: string | null;
  vintage: number | null;
  grape_variety: string[] | null;
  region: string | null;
  country: string | null;
  type: WineType;
  taste_profile?: TasteProfile;
  food_pairing?: string[];
  flavor_notes?: string[];
  description?: string;
}

export interface ScanResult {
  scan_id: string;
  confidence: number;
  wine: ScannedWine;
  image_url: string;
  existing_wine_id: string | null;
  is_duplicate: boolean;
}

export interface BatchScanResultItem {
  index: number;
  status: 'success' | 'failed';
  confidence: number | null;
  wine: ScannedWine | null;
  bounding_box: { x: number; y: number; width: number; height: number } | null;
  error: string | null;
}

export interface BatchScanResult {
  scan_session_id: string;
  total_detected: number;
  successfully_recognized: number;
  failed: number;
  wines: BatchScanResultItem[];
}

// Recommendation types
export interface RecommendationItem {
  rank: number;
  match_score: number;
  user_wine: UserWineListItem;
  reason: string;
  pairing_tips: string | null;
  drinking_urgency: DrinkingUrgency;
}

export interface RecommendationResult {
  recommendation_id: string;
  query: string;
  recommendations: RecommendationItem[];
  general_advice: string | null;
  no_match_alternatives: string | null;
  created_at: string;
}

export interface RecommendationRequest {
  query: string;
  query_type?: 'food' | 'occasion' | 'mood';
  preferences?: {
    wine_types?: WineType[];
    max_results?: number;
    prioritize_expiring?: boolean;
  };
}

// AI Analysis types
export interface WineAromaProfile {
  primary: string[];
  secondary: string[];
  tertiary: string[];
}

export interface WineAgingPotential {
  current_status: string;
  recommendation: string;
  peak_window: string;
}

export interface WineFoodPairingDetail {
  dish: string;
  reason: string;
}

export interface WineVivinoRating {
  estimated_score: number;
  confidence: string;
  note: string;
}

export interface WineAIAnalysis {
  summary: string;
  aroma_profile: WineAromaProfile | null;
  flavor_analysis: string | null;
  terroir_context: string | null;
  aging_potential: WineAgingPotential | null;
  food_pairing_detail: WineFoodPairingDetail[] | null;
  sommelier_tip: string | null;
  vivino_rating: WineVivinoRating | null;
  comparable_wines: string[] | null;
}

// Dashboard types
export interface CellarSummary {
  total_wines: number;
  total_bottles: number;
  total_value: number;
  by_type: Record<WineType, { count: number; bottles: number }>;
  by_country: Array<{ country: string; count: number }>;
  by_cellar: Array<{ tag_id: string; name: string; count: number }>;
  recent_additions: number;
  consumed_this_month: number;
}

export interface ExpiringWineItem {
  user_wine_id: string;
  wine: {
    name: string;
    vintage: number | null;
  };
  drinking_window_end: number;
  years_left: number;
  message: string;
}

export interface ExpiringWines {
  urgent: ExpiringWineItem[];
  soon: ExpiringWineItem[];
  optimal: ExpiringWineItem[];
}

// Tag types
export interface TagCreateRequest {
  name: string;
  type: TagType;
  color?: string;
}

export interface TagUpdateRequest {
  name?: string;
  color?: string;
}

export interface TagListResponse {
  tags: Tag[];
  summary: {
    cellar_count: number;
    location_count: number;
    custom_count: number;
  };
}
